### ISSUE TITLE
What we're exploring with FatBeacon

### ISSUE THREAD

**Issue by scottjenson** [google/physical-web#784]
The latest release of the Physical Web app has support for not only wifi-direct and mDNS, but also a new type of beacon, we're calling 'fatbeacon'. The idea is to still broadcast an Eddystone-URL packet but it just contains the title of the beacon. If the user selects it, the app connects using a GATT service, pulls the html file out and then opens it in Chrome.

We call this the 'cabin in the woods scenario' as you can walk up and interact with a fatbeacon even if you don't have any data.  At the moment, there is unfortunately no way for you to test it. However, we're going to be releasing a new version of the Physical Web app which will broadcast as a fatbeacon. That should be coming in just over a week. 

What we'd like this experiment to do though is to kick off a discussion. There are a few big topics that need to be discussed a bit more:
* How do we handle security when there is no way to vet the beacon or it's contents?
* The data size of a page using BLE is likely going to be very small (<40k) Are people willing to work within such a small footprint? In our initial tests, it's possible to download about 10k/sec over existing BLE beacons. While this will get a bit better with BLE 5, it will never be allow big pages to load.
* It's a single page HTML file so no links to images, etc. This is very restrictive for the dev

Most of these issues are well above the specifics of our implementation today. In fact, we need to stress that everything we're doing with fatbeacon could be thrown away. We're just trying to kickstart this discussion with an early build. As we learn more, we're happy to change directions.

**Comment by hardillb** [comment]
A few thoughts off the top of my head

 1. Images can be base64 inline encoded (but that's going to bloat things very quickly)
 2. Any thoughts on gzip compressing (like the http servers do) which could allow you to cram more into your 40k? 
 3. If you broadcast a URL could this be a fall back if data access is not available?
 4. Since this is going to have to run in it's own webview in the app (as I can't see a way to pass it directly to Chrome just yet) could you add a tighter sandbox on what the page can do to help with security?

**Comment by scottjenson** [comment]
1) Agreed
2) Seems like a good idea
3) Just broadcast 2 URLs one fatbeacon and the other a normal URL (of course, it would be good if the scanner could show one or the other)
4) Actually, we're just downloading the file and calling Chrome, it opens right up

**Comment by jokeyrhyme** [comment]
Just a thought: the HTML payload doesn't necessarily need to be stored on the beacon. The beacon could just be serving as a Bluetooth proxy. Handy when the "beacon" has a hardline, and cellular reception on the client device is spotty or unavailable.

**Comment by scottjenson** [comment]
But that's why we're also considering wifi-direct. The device could be hardwired and up to date, using wifi-direct for passers by. The other advantage to wifi-direct is that it can server more users and is much faster.

**Comment by ferencbrachmann** [comment]
@scottjenson What is the goal with goal of fatbeacon?

We talked it over and we have a suggestion: why are you puting the HTML on the fatbeacon? Instead we suggest looking at puting individual service URLs instead of the limited HTML on there.
This URL could point to content on the beacon or preferably to a remote host. It also looks a LOT safer if it's only a remote content URL on there as HTML can open the Pandora's box of injection attacks.

**Comment by scottjenson** [comment]
@ferencbrachmann the goal is to work without any data connectivity. I'm not sure what you mean by "individual service URLs" but any URL would fail if you have no connectivity.

**Comment by philips77** [comment]
Hi, very interesting idea, those fat beacons. My thoughts:
If you have no Internet, you usually don't expect super fancy websites. You don't expect them at all. Small page with some CSS and some buttons to control a device with WebBluetooth should easily fit requirements. It would have to be built in Chrome (connecting and obtaining HTML) when adopted, like it supports Physical Web now.
As the beacon would have to be connectible, you still have Scan Response to put more data, but what is the second URL you're talking about? Shouldn't it just be Service UUID that this service is supported + a complete local name to display on the notification?

**Comment by jokeyrhyme** [comment]
One benefit of the cloud registry look-up was that Google could help prioritise the list of beacons, and could remove malicious ones from the list. While I love the idea of a cloud-free web page, I do wonder about security, etc.

Could an attacker have their own copy of a vending machine's beacon?
What mechanisms are in place to ensure that users connect to the correct beacon?
Does HTTPS even help us here?
Would the vending machine company want a private certificate key to be physically installed on the beacon?

**Comment by scottjenson** [comment]
I'm assuming your comments are directed to fatbeacons? If so, HTTPS doesn't apply as there is no URL to speak of.

Security is clearly a big concern. We agree the a cert is likely the best way to go. We have to figure out a way to incorporate that into the web page so we can be sure it's coming from a valid domain. This could be our variation of HTTPS...

**Comment by GregCornelius** [comment]
Hey Scott. We have been down this track already working on a project for our local Tourism Board who wanted to get content to tourists who are travelling without any data connectivity. We built a prototype with a Raspberry Pi behaving like a beacon. We formatted the Pi as web server and host. As you know Pi 3 has both BLE and Wi-Fi. Our user testing revealed that most tourists do not have Bluetooth turned on, and are not so familiar with it, but they do have Wi-Fi turned on and are familiar with it (so I'm pleased you are looking at Wi-Fi direct). The next obstacle was the captive portal on the phones. JavaScript is disabled on captive portals on iOS so the web page has to be really dull - text and image only with no external actions. There isn't much value for the user there. On Android it's not so bad. Anyway, we struggled to establish a sufficient use case. When you combine the security risk with that fact that the majority of people do have data plans, and many places offer free Wi-Fi for the remainder, then the opportunity is very very small. In our case the Tourism Board decided not to proceed any further with the project. To be honest we have stopped working with BLE beacons now and instead started working with a special Wi-Fi router and re-directs to our target webpage after the user selects the SSID (or after log-in). Our testing has found easily more than 100 times as many people choose the Wi-Fi log in option compared to the BLE beacon (when we offer both). To be clear, for every 1 beacon log-in we are getting more than 100 Wi-Fi log ins....

**Comment by ferencbrachmann** [comment]
@scottjenson Could you amplify on your goals with the fatbeacon at the use-case level?

**Comment by scottjenson** [comment]
Our extreme case is "Cabin in the woods" where you use a beacon in the middle of no where to open a web page and then unlock the door. That's clearly extreme but it makes the point that no connectivity is required.  The main reason we're looking at this is developing countries where coverage is spotty and or expensive. Ultimately we see phones as being possibly the biggest 'fatbeacons' as you could use a small app on your phone to curate the information you want and then share it will anyone nearby. The obvious case here would be simple store fronts.
 
But there appear to be use cases for countries with coverage, so games in the car, toys for kids, or controlling a car stereo all seem like use cases that could benefit from working under limited connectivity conditions. 

Keep in mind that we have been pummeled by people that wanted a way to use the Physical Web without connectivity. Fatbeaacon is a tiny step in exploring that direction (so is WifiDirect). We don't expect this to be easy and this initial direction clearly has problems. We understand that. However, part of working this way is that we try something out and share it openly. We've been amazed how the insightful and helpful the Physical Web community has been.

**Comment by ferencbrachmann** [comment]
I would imagine this will have relevance in India/Africa and some parts of Asia. 

We had some landing page performance issues with our Beeem service with 2G connections mainly in India. We're doing basic research into PWAs and in a few months all of our Beeem landing pages are going to be Progressive Web Apps allowing for excellent performance even in 2G and off-line functionality aswell (of course this still requires network connectivity to download the PWA in the first place).

The concern with security is that if you open the avenue for onboarding then a lot of attack types become possible that might threaten the entire physical web concept.

**Comment by GregCornelius** [comment]
I have experience of issues like these in emerging Asia. For example a major Indonesian hotelier who moved his entire business to the cloud. His staff had to resort to pen and paper transactions because it's not just the internet that goes down; it's also electricity. The cloud is a disaster in some parts of the world. However, what would be the role of a fat beacon in this situation? It's only good if you can read and write. A better solution is a local PC with battery back up, that processes transactions locally, and then updates the cloud storage when connectivity / electricity is restored. This is edge computing AKA the fog. I guess a FatBeacon could be part of fog network along with other components... I can see some value in a FatBeacon helping to reduce the demand on a slow network perhaps? The other thing to consider is that over time the telcos are solving the bandwidth issues in emerging nations. This wont be a problem forever. 

How is a FatBeacon concept better than a R Pi, say? I can think of battery life. 2+ years with a beacon. About 24 hours with Pi. Any other advantages?

**Comment by angst7** [comment]
Battery life is a pretty significant advantage.  Imagine a device out in the woods that marks a trail, and attaches basic direction/distance information.  You don't need fancy markup, or javascript.  You do need it to do its job unattended for a long time.

**Comment by jokeyrhyme** [comment]
@GregCornelius 

> How is a FatBeacon concept better than a R Pi, say?

You probably mean: "how is a purpose-built battery-optimised FatBeacon implementation better than an implementation that uses a Raspberry Pi?"

FatBeacon is more of a concept and a proposal / specification, and the Raspberry Pi is physical hardware that already exists. They don't really compete with each other, and there is definitely opportunity here for them to complement each other.

**Comment by jokeyrhyme** [comment]
@GregCornelius 

Alternatively: "how is a FatBeacon implementation (BLE / WiFi Direct) better than a Raspberry Pi acting as a WiFi access point / captive portal with a default web page payload?"

The benefit is that FatBeacon as a specification can be implemented by device manufacturers, handset vendors, and smartphone OS providers. There's opportunity here for safer and more specific user experience to be offered.

**Comment by scottjenson** [comment]
That is why we're doing both. Wifi Direct is likely plugged into a wall but has a huge advantage in that it can service multiple users at once. The fatbeacon may be more helpful running as a phone client allowing anyone to create a page on their phone and share it to anyone nearby.

The 'fatbeacon on a path in the woods' is likely a tough problem. A beacon running on 2AA cells could last many years. Connecting and downloading a web page would certainly take significant battery but it would still likely last quite a while. If you would add a PV cell to recharge it, it could last indefinitely. But lets be clear, that's likely not the first use of this technology.

**Comment by beaufortfrancois** [comment]
For info, I've created a simple web app that reads FatBeacon content at https://beaufortfrancois.github.io/sandbox/physical-web/fatbeacon/
Let me know what you think.

**Comment by hardillb** [comment]
Check out issue #814 looks like we have Node.js and python implementations for people to try

**Comment by jokeyrhyme** [comment]
There are some aspects of FatBeacon that overlap somewhat with Mozilla's FlyWeb:

- https://hacks.mozilla.org/2016/09/flyweb-pure-web-cross-device-interaction/
- https://flyweb.github.io/
- https://github.com/flyweb/spec

Has there been any collaboration between Google and Mozilla teams in this area? Is that something that is worth considering?

**Comment by scottjenson** [comment]
We have had discussions with the flyweb team and are happy to continue to do so. However, the approaches appear to be quite different so it's not clear how they could collaborate. Do you have any specific proposal? We're happy to discuss it.

**Comment by bitsoko-services** [comment]
we are currently using ble beacons for kiosk merchants in kenya. we have found the service worker to be indispensable when using beacons. in our case, the fatbeacons could used to store only unique strings/blobs (instead of a webpage) whose data the service worker would use in a template to produce a webpage.

**Comment by angst7** [comment]
@sandeepmistry I agree, though I don't know if this is the right place to discuss it.  I brought it up in #791 and proposed an alternate solution involving multiple characteristics, but I really don't like that solution either.  I'm particularly interested in thinking through how to write/update the fatbeacon content.  The spec doesn't address that, and there should be some GATT friendly, standardized way to do it.  

Perhaps start a new issue to consolidate this discussion, if other folks think it worth talking through.

**Comment by jacobrosenthal** [comment]
Related, You guys have a good contact for android ble? The mtu change followed by reads is coming into osx as a read, then mtu change, then more reads...This is breaking the fatbeacon logic for last packet as android thinks it had a successful mtu change BEFORE any reads started. Moving to something more standard like notifications/indications would also fix.

**Comment by gfwilliams** [comment]
Just stumbled upon this - it looks really exciting, and I'd love to make sure Espruino/Puck.js is able to handle it when it's released.

Some thoughts:

* 40k is enough to do some amazing stuff. js1k has some good examples :) Worst case the HTML could drag CSS off a server, so you still get a usable page without internet.
* Gzip seems like a good plan, but should be optional - some beacons may want to send data that changes, and may not have room for a gzip encoder.
* The whole repeated read thing seems clunky - +1 for notifications or something else.
* Reading the content length in the first bytes would allow you to have a progress bar
* Would a connection be required to read? If not, the HTML data could actually be broadcast interleaved with the advertising, like is done for eddystone with other data. That way, the HTML might even be loaded onto the phone before the user even clicked the notification, making it much faster. There might be battery life considerations for beacons though.
* IMO the ability to write HTML content should be something the beacon manufacturer figures out - there's no need for it to be part of the spec.
* I don't see how this is less secure than going to a malicious website or loading a malicious HTML file. I'm not sure any more sandboxing than usual is required?
* Given it looks like there's not going to be Web Bluetooth support for iOS, two way communications could be difficult. Perhaps the webpage that data is loaded into could expose a pre-connected Web Bluetooth object that provided the ability to change characteristics *on that device only*.
* Or, you could connect, open a Nordic UART-style connection where the beacon pushes content length + webpage via notifications, and then stay connected and provide an open websocket connection to the beacon that worked by reading and writing the 2 characteristics?
* Even if read only I see this being super useful for stuff like dataloggers. It'd be easy to just walk up to one, click the notification, and see the last day's data.

**Comment by bluetooth-mdw** [comment]
Bluetooth 5 gives 8 x larger advertising packets so it might make it possible to do this in a connectionless mode rather than requiring a connection, albeit still with significant constraints. You'll also have a problem in that despite the fact that there's no specified limit to the number of concurrent connections a device can accept, in the real world there always is and it's usually a surprisingly small number. Apologies if these points have already been made... this thread was just brought to my attention and I thought it looked interesting!

**Comment by scottjenson** [comment]
Interesting point. However, in order to do this, it would require a large quantity of numbered packets. A 20K HTML file would require at least 40 max size packets. In order to make that come across quickly enough, the adv frequency would need to be increased significantly, likely to 100m which would then take 4 seconds to get it all. If any packets were missed, it would require another 4 seconds to fill in the blanks. Given that you'd still want to send the Title packet as well, that might have to be interleaved in there, making it a bit slower. Given that the normal adv frequency of Eddystone-URL is 700ms this would drain the battery roughly 7x faster as well.

I completely appreciate the value of doing this connectionless by the way. I'm raising these issues to explore possible solutions. Do you think there are work arounds to the problems I proposed?

**Comment by gfwilliams** [comment]
Maybe there's a way to poke it and say 'hey, there's someone here. Can you speed up advertising for the next few seconds' ?

It doesn't solve the missed packet issue though.

**Comment by bluetooth-mdw** [comment]
Hi Scott, 20K is pretty large in the energy efficient, battery aware world of Bluetooth low energy as you know and so, no, if that's the goal then a connectionless approach doesn't look a good fit for you. 

Therefore you'll be faced with the concurrent connections limit that commercially available modules typically have. This may or may not be an actual problem of course. It may well be fine. Depends on the use case.

Unless (sound of mental cogs whirring)... there's a hybrid approach that might work? What follows here may be nonsense but I'll take the risk..... what if you were to broadcast the most fundamental content.... the raw HTML for example and then for the client to download referenced content like BASE64 encoded images, over a connection? The user experience would be like that of a 1980s modem perhaps (reminisces) with the full page downloading in stages.....

**Comment by scottjenson** [comment]
Not silly at all, we actually had the same thought: download the JS that would open a connection and get what you need. Much more complex though and we expect unlikely to be used by many folks.

We are assuming fairly low usage so the number of connections should be survivable. If you need high numbers of connections, that's why we've also explored Wifi-Direct.

The whole point of this release is to have these types of conversations so if there are whirring ideas, please do share) ATM though, I'll cautiously suggest that the connectionless approach doesn't seem practical for any non-trivial page size.

**Comment by bluetooth-mdw** [comment]
Yep, I have to agree. Useful to kick ideas around though, as you say (no matter how crazy!).

If you arrive at a point where you think it's appropriate to discuss 'standards' (e.g. for the profile which yields access to the HTML on the device), please contact me. It would be good to consider formally adopting whatever you come up with here I think. Looks very interesting to me.

**Comment by gfwilliams** [comment]
> the profile which yields access to the HTML on the device

A standard BLE UUID for basic file access (write filename to a characteristic, read file from another) would be amazing.

**Comment by bluetooth-mdw** [comment]
I'm thinking that sounds like a standard serial data comms profile, rather than 'file access' specifically. I may be wrong and don't want to take us off topic here. FYI anyone can propose a new standard profile to the Bluetooth SIG.

**Comment by scottjenson** [comment]
You can actually read and write to the same characteristic. If there is an existing BLE standard we should use, that would clearly be preferred.

**Comment by dennisg** [comment]
Interesting thread, this is...

Regarding the reasonable limit on the number of bytes to transfer there's also the approach where the spec could define a set of 'shell' (or template) html pages that only need a few items filled in.

In the end we might be able to conclude that a lot of pages FatBeacons serve look quite similar,  
and we only need to supply e.g. only the title, some introductory text, a question and a set of possible answers (up or down, open or close, left or right, yes or no). Just like e.g. the Chromecasts's Styled Media Receiver, most of the code is already there, only the styling is altered (a subset of all required data)

With a single byte we would be able to differentiate between more than 255 (;-)) pre-defined pages that will have been pre-loaded on the phone, leaving only a small subset of the actual data that needs to be downloaded over BLE.

Given what @bluetooth-mdw mentions about BLE 5 adv packet sizes this data might even fit in one or two new advertisements. Awesome if this would be accomplishable connectionlessly...

Regarding the requirement, currently, to only load a single page: It seems to me like a real improvement if we could somehow load additional data (using another (set of) characteristics maybe?) to fulfill external requirements (css/ js/ images). It is probably possible to mimic full-blown HTTP/1.0 over BLE but if that's sensible is another matter.

altogether: awesome stuff!

**Comment by adriancretu** [comment]
I think HTML/JS is a terrible overhead given BLE transfer speeds of 1.5 kB/s for iOS. I wouldn't be surprised if a new "beacon app" byte-code / protocol would be invented. Unfortunately today's best JS frameworks generate a production bundle for a basically empty "hello world" webapp of ~300 kB (gzip-ed and minified!).  And even an old-school hand-written .JS file would end-up at a few KB.

**Comment by scottjenson** [comment]
Yeah, this discussion is *exactly* why we pushed this out. We need to make sure we look at all options and even weigh the viable of the entire concept.

@dennisg wrt your template idea. So many companies have gone down this route and it seems to always end in a mess. Too many folks fighting over their template on one end and the extremely limited use options offered.

Most folks I talk to are very excited by the possibility of getting WebBT in the page which would allow for a "Happy Mean Toy' Scenario, e.g. controlling a very inexpensive toy. I don't see how that would be possible with fixed page layouts (but I'm happy for you to point out how it could)

@adriancretu I appreciate you point but BLE5 will support MUCH higher transfer rates so I hope this will get much better soon. At to your issue about waiting for apple to support it, I agree, that's a problem, but waiting for Apple effectively seeds control to them. We need to figure out what needs to be done and do it. I'm don't want to live in a world where the only way forward is to ask Apple's permission. Besides, it's not likely going to be that bad. BLE5 offers many new advantages and it seems reasonable to assume that Apple will be improving their transfer speeds as they support BLE5. It may not happen in a few weeks, but it seems inevitable.

**Comment by smartyw** [comment]
@adriancretu @scottjenson Actually, 4.2 supports much higher data rates! But yes, Bluetooth 5 doubles data rates again, compared with 4.2.

https://www.bluetooth.com/~/media/files/specification/bluetooth-quic-ref-guide.ashx?la=en

**Comment by gfwilliams** [comment]
You can do useful sites with very little HTML+JS. It's not like you need to pull in jQuery to deal with browser quirks any more. I'm not sure special bytecode would be much better than minified+gzipped HTML+JS either - at least it wasn't in my tests.

For example [this UI](http://www.espruino.com/Interactive+Web+UI) was a quick hack and isn't made for size, but minified and gzipped it's still under 1700 bytes.

Sure, you may have to design a webpage with FatBeacon in mind a little, but that's a small price to pay for the benefits.

**Comment by dennisg** [comment]
@scottjenson yes it would be very difficult to please everybody, maybe not the right track indeed. 

Okay, thinking out loud here: another option would be to pre-define a (limited) list of polymer-elements and make that list public for reference. Those should then cached for offline availability. Some kind of hook to load these html imports would be required, I guess...

If we include the webBT-related polymer-elements, access through BLE to a low-cost device would seem feasible to me with minimal HTML+JS.

In time, of course, the list of polymer-elements can be extended upon. Probably based on requests and popular demand. 

I can try to cook up an example where I read e.g. the temperature from a SensorTag I have lying around if you want?

**Comment by Timmmm** [comment]
Probably an obvious question, but I assume WebBluetooth is enabled for these pages, even though they aren't loaded through HTTPS? 

Also I can't see it explicitly stated but does this actually work now? As in, if I make a FatBeacon can normal Android users see it in the same way they can see Eddystone-URL beacons? Or does it only work through the Physical Web app?

Oh and where is the spec? (Or at least a brief description of the GATT service so I can implement it?)

**Comment by hardillb** [comment]
@Timmmm FatBeacons currently only show up in the PhysicalWeb app (because this is currently just an experiment so not likely to make it into production core OS code just yet).

There is a description of the GATT details in issue #814 and there are a few implementations listed in there that you can have a look at such as the one in the PhysicalWeb app code, My NodeJS code and a Python version.

The html is actually loaded as a local file:// url. I need to actually test if WebBluetooth will work from there but it will need the flag setting as I don't think you'll be able to get a Origin-Trial token for file://

**Comment by angst7** [comment]
@Timmmm there's additional discussion of the spec in #791.  There's also a link in the comments of that issue to an nRF52 implementation of fatbeacon.

**Comment by scottjenson** [comment]
@sandeepmistry sorry I missed your comment in this thread. It might be better to pull this into another issue. My understanding is that the fatbeacon is a very simplistic (but hopefully) GATT like service. It's effectively doing a 'streamed read/write' to a characteristic, reading or writing multiple times from the same characteristic. I was under the impression that many GATT services would do this. Am I mistaken?

**Comment by ukBaz** [comment]
@scottjenson I've been on the BlueZ mailing list getting some tips on how to improve the implementation of my fatbeacon on a Linux Single Board Computer. As part of the discussion there were a couple of bits of information that I thought were worth sharing here as they seem to agree with what @sandeepmistry has said about implementing fatbeacon over GATT.
Firstly I got:
> Since a characteristic can only be 512 bytes, you can never use Read Requests or Read Blob Requests to get more than 512 bytes if the characteristic is not updated. Normally when you want to stream over a lot of data over BLE you don't put everything in a GATT db and then read it. Instead you use set up to send a lot of notifications for example when you receive some write to a control point. Since there is no limitation how many notifications you can send, just fragment your data into multiple notifications (according to the ATT MTU size) and send them over.

Luiz followed up with:
> As you may have guessed this [current implementation] is non-standard and should probably be changed to use notification as pointed out by Emil

**Comment by bluetooth-mdw** [comment]
It seems to me, what you want is something which emulates serial communications, sitting on top of GATT. There are custom services that do just that, for example the UART service from Nordic Semiconductor:

https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk51.v9.0.0%2Fgroup__ble__sdk__srv__nus.html 

Notifications are used by this service, as described by @ukBaz

**Comment by scottjenson** [comment]
In general, we've always taken the 'try to be like the web' approach to what we do, so your suggestion is entirely reasonable.

We would love for the BT SIG to bless an HTTP/Serial approach that we'd just use directly. There seems to be some movement in that direction but my guess is that it's just an obvious direction to take given how slow any implementation is likely to be.

The FatBeacon approach was just an experiment to encourage questions like this. It was tackling the slow speed issue by creating this single point of transfer. It's clearly not the like web at all, but it does force the implementer to compress everything into a single, much likely significantly smaller, packet. My worry (which may be unfounded) is that a serial/http approach will be so excruciatingly slow that it just won't be practical.
